libs/payments/ui/src/lib/nestapp/nextjs-actions.service.ts

```diff
@@ -253,6 +256,95 @@ export class NextJSActionsService {
     return data;
   }
 
+  @SanitizeExceptions()
+  @NextIOValidator(
+    DetermineStaySubscribedEligibilityActionArgs,
+    DetermineStaySubscribedEligibilityActionResult
+  )
+  @WithTypeCachableAsyncLocalStorage()
+  @CaptureTimingWithStatsD()
+  async determineStaySubscribedEligibility(args: {
+    uid: string;
+    subscriptionId: string;
+    acceptLanguage?: string | null;
+    selectedLanguage?: string;
+  }) {
+    try {
+      const cmsChurnResult =
+        await this.productConfigurationManager.getChurnInterventionBySubscription(
+          args.subscriptionId,
+          'stay_subscribed',
+          args.acceptLanguage || undefined,
+          args.selectedLanguage
+        );
+
+      const cmsChurnInterventionEntries = cmsChurnResult.getTransformedChurnInterventionByProductId();
+      if (!cmsChurnInterventionEntries.length) {
+        return {
+          isEligible: false,
+          reason: 'no_churn_intervention_found',
+          cmsChurnInterventionEntry: null,
+        }
+      }
+
+      const cmsChurnInterventionEntry = cmsChurnInterventionEntries[0];
+      let redemptionCount = 0;
+      try {
+        const churnInterventionEntryData = await this.churnInterventionManager.getEntry(
+          args.uid,
+          cmsChurnInterventionEntry.churnInterventionId
+        );
+        redemptionCount = churnInterventionEntryData.redemptionCount ?? 0;
+      } catch (error) {
+        if (error instanceof ChurnInterventionEntryNotFoundError) {
+          redemptionCount = 0;
+        } else {
+          throw error
+        }
+      }
+
+      if (redemptionCount >= cmsChurnInterventionEntry.redemptionLimit) {
```

Comment:
Note: `redemptionLimit` is not a required field, as RPs may want to allow customers to redeem a coupon indefinitely.